### PR TITLE
Fix/UI kit 3053/modify issues on voice message

### PR DESCRIPTION
**Pull request by HoonBaek**
[UIKIT-3256](https://sendbird.atlassian.net/browse/UIKIT-3256)

## Description Of Changes

Creates

* UI components
  ```javascript
  import PlaybackTime from "@sendbird/uikit-react/ui/PlaybackTime"
  import ProgressBar from "@sendbird/uikit-react/ui/ProgressBar"
  import VoiceMessageInput from "@sendbird/uikit-react/ui/VoiceMessageInput"
  import VoiceMessageItemBody from "@sendbird/uikit-react/ui/VoiceMessageItemBody"
  ```
  * PlaybackTime: Display the current time in 00:00 format with the received millisecond value
  * ProgressBar: Display the current progress status with the received maxSize and currentSize of millisecond unit value
  * VoiceMessageInput: UI component for recording and playing a voice message
  * VoiceMessageItemBody: UI component for rendering a voice message also able to play voice message
* VoiceRecorder
  ```javascript
  import { VoiceRecorderProvider, useVoiceRecorderContext } from '@sendbird/uikit-react/VoiceRecorder/context'
  import useVoiceRecorder from '@sendbird/uikit-react/VoiceRecorder/useVoiceRecorder'
  ```
  * VoiceRecorderProvider: A react context provider component providing `start`, and `stop` functions
  * useVoiceRecorderContext: A react useContext hook of VoiceRecorderProvider
  * useVoiceRecorder: A react hook that provides advanced context, `recordingLimit`, `recordingTime`, `recordingFile`, and `recordingStatus`. Recommend using this hook in the customized components.
* VoicePlayer
  ```javascript
  import { VoicePlayerProvider, useVoicePlayerContext } from '@sendbird/uikit-react/VoicePlayer/context'
  import useVoicePlayer from '@sendbird/uikit-react/VoicePlayer/useVoicePlayer'
  ```
  * VoicePlayerProvider: A react context provider component providing `play`, and `pause` functions
  * useVoicePlayerContext: A react useContext hook of VoicePlayerProvider
  * useVoicePlayer: A react hook that provides advanced context, `playbackTime`, `duration`, and `playingStatus`. Recommend using this hook in the customized components.
* utils/isVoiceMessage: A function that you can check if the given message is a voice message
  ```javascript
  import isVoiceMessage from '@sendbird/uikit-react/utils/message/isVoiceMessage'
  const isVoiceMsg: boolean = isVoiceMessage(message);
  ```

Edits

* Install `lamejs` to transcode the audio file to mp3
* Add props `isVoiceMessageEnabled` and `voiceRecord` props to the App, `SendbirdProvider`, and `MessageInput` components, to turn on/off the voice message recording feature
  ```javascript
  <SendbirdProvider
    isVoiceMessageEnabled
    voiceRecord={{
      maxRecordingTime: 60000,
      minRecordingTime: 1000,
    }}
  >
    {/* implement custom application */}
  </SendbirdProvider>
  ```
* Add props `renderFileUploadIcon`, `renderVoiceMessageIcon`, and `renderSendMessageIcon` into the `Channel`, `ChannelUI`, and `MessageInput` component
* Add `onVoiceMessageIconClick` props to the `MessageInput` component
* Add `onBeforeSendVoiceMessage` props to the `Channel` component
* Fetch message list including `MetaArray` in the `Channel` and `Thread` modules
* Provide new icon-type `AudioOnLined` & new icon-color Primary2 and OnBackground4
* Provide new string sets
  ```javascript
  <SendbirdProvider
    stringSet={{
      BUTTON__OK: 'OK',
      VOICE_MESSAGE: 'Voice Message',
      MODAL__VOICE_MESSAGE_INPUT_DISABLED__TITLE_MUTED: 'You\'re muted by the operator.',
      MODAL__VOICE_MESSAGE_INPUT_DISABLED__TITLE_FROZEN: 'Channel is frozen.',
    }}
  >
    {/* implement custom application */}
  </SendbirdProvider>
  ```
  * `BUTTON__OK`: 'OK' → Used on the submit button of pop up modal
  * `MODAL__VOICE_MESSAGE_INPUT_DISABLED__TITLE_MUTED`: 'You\'re muted by the operator.' → Used in an alert pop-up modal
  * `MODAL__VOICE_MESSAGE_INPUT_DISABLED__TITLE_FROZEN`: 'Channel is frozen.' → Used in an alert pop-up modal
  * `VOICE_MESSAGE`: 'Voice Message' → Used in ChannelPreviewItem, QuoteMessage, and MessageSearch to appear that the message type is the voice

## Types Of Changes

What types of changes does your code introduce to this project?
Put an `x` in the boxes that apply_

- [ ] Bugfix
- [x] New feature
- [ ] Documentation (correction or otherwise)
- [ ] Cosmetics (whitespace, appearance (ex) Prettier)
- [ ] Build configuration
- [ ] Improvement (refactor code)

[UIKIT-3256]: https://sendbird.atlassian.net/browse/UIKIT-3256?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -52,6 +51,7 @@
     "@sendbird/chat": "^4.2.3",
     "css-vars-ponyfill": "^2.3.2",
     "date-fns": "^2.16.1",
+    "lamejs": "^1.2.1",
```
Comment (sravan-s) [inline package.json:54]: To add lame
```
1. Link to LAME as separate jar (lame.min.js or lame.all.js) 

> for this, add lamejs to rollup.externals -> https://github.com/sendbird/sendbird-uikit-react/blob/main/rollup.config.js#L40
https://rollupjs.org/configuration-options/#external

2. Fully acknowledge that you are using LAME, and give a link
   to our web site, lame.sourceforge.net
   
   > Add lameJS to our readme

3. If you make modifications to LAME, you *must* release these
   these modifications back to the LAME project, under the LGPL.
   
> not applicable
```

File: src/hooks/VoicePlayer/voicePlayerEvent.ts [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+  onPlayingStopped?: (props: VoicePlayerEventParams) => void;
+  onPlaybackTimeUpdated?: (props: VoicePlayerEventParams) => void;
+}
+export interface VoicePlayerEventParams {
+  groupKey: string;
+  duration?: number;
+  playbackTime?: number;
+}
+
+export const generateGroupKey = (channelUrl: string, key: string): string => (`${channelUrl}/${key}`);
```
Comment (sravan-s) [inline src/hooks/VoicePlayer/voicePlayerEvent.ts:38]: Add newline to end of file pls

File: src/hooks/VoiceRecorder/WebAudioUtils.ts [hunk 3]
```diff
@@ -0,0 +1,134 @@
+// Thanks to https://codesandbox.io/s/media-recorder-api-downsampling-16k-mp3-encode-using-lame-js-forked-n1pblw
```
Comment (sravan-s) [inline src/hooks/VoiceRecorder/WebAudioUtils.ts:1]: 👍

File: src/hooks/VoicePlayer/index.tsx [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+});
+
+export const VoicePlayerProvider = ({
+  children,
+}: VoicePlayerProps): React.ReactElement => {
+  const [eventStorage, setEventStorage] = useState<VoicePlayerEventStorage>({});
+  const [currentPlayer, setAudioPlayer] = useState<HTMLAudioElement>(null);
+  const [currentGroupKey, setCurrentGroupKey] = useState<string>('');
+
+  const addEventHandler = (event: VoicePlayerEvent): void => {
```
Comment (sravan-s) [inline src/hooks/VoicePlayer/index.tsx:41]: was better to have event_handler as a seperate system, and add tests to it ~
But if you can add tests to voice player, its good enough

File: src/hooks/VoicePlayer/useVoicePlayer.tsx [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+          onPlayingStarted(props);
+          setDuration(props?.duration);
+          setPlaybackTime(props?.playbackTime);
+        },
+        onPlayingStopped: (props) => {
+          const { duration, playbackTime } = props;
+          setPlayingStatus(VoicePlayerStatus.READY_TO_PLAY);
+          onPlayingStopped(props);
+          setDuration(props?.duration);
+          if (duration - playbackTime <= 0.01) {
```
Comment (sravan-s) [inline src/hooks/VoicePlayer/useVoicePlayer.tsx:70]: `0.01 ` -> can we add const for this

File: src/hooks/VoiceRecorder/index.tsx [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+    </VoiceRecorderContext.Provider>
+  )
+};
+
+export const useVoiceRecorderContext = (): VoiceRecorderContext => useContext(VoiceRecorderContext);
+
+export default {
+  VoiceRecorderProvider,
+  useVoiceRecorderContext,
+};
```
Comment (sravan-s) [inline src/hooks/VoiceRecorder/index.tsx:101]: Make a ticket to add logger to Voice player & recorder ~ not a blocker

File: src/hooks/VoiceRecorder/useVoiceRecorder.tsx [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+    stopTimer();
+  }, [voiceRecorder]);
+  const cancel = useCallback(() => {
+    stop();
+    setRecordedFile(null);
+  }, [voiceRecorder]);
+
+  // Timer
+  const [recordingTime, setRecordingTime] = useState<number>(0);
+  let timer: NodeJS.Timer = null;
```
Comment (sravan-s) [inline src/hooks/VoiceRecorder/useVoiceRecorder.tsx:74]: NodeJS.Timer -> might cause type issue? no?

Comment (sravan-s) [reply src/hooks/VoiceRecorder/useVoiceRecorder.tsx:74]: https://stackoverflow.com/a/59681620 is better

File: src/smart-components/Channel/components/MessageInput/VoiceMessageInputWrapper.tsx [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+        }}
+        onControlClick={(type) => {
+          switch (type) {
+            case VoiceMessageInputStatus.READY_TO_RECORD: {
+              pause();
+              start();
+              break;
+            }
+            case VoiceMessageInputStatus.RECORDING: {
+              if (recordingTime >= 1000 && !isDisabled) {
```
Comment (sravan-s) [inline src/smart-components/Channel/components/MessageInput/VoiceMessageInputWrapper.tsx:110]: recordingTime >= 1000 > better to be const

File: src/smart-components/Channel/context/hooks/useSendVoiceMessageCallback.ts [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+  }: StaticParams): Array<FuncType> => {
+  const sendMessage = useCallback((file: File, duration: number, quoteMessage: UserMessage | FileMessage) => {
+    const messageParams: FileMessageCreateParams = (
+      onBeforeSendVoiceMessage
+      && typeof onBeforeSendVoiceMessage === 'function'
+    )
+      ? onBeforeSendVoiceMessage(file, quoteMessage)
+      : {
+        file,
+        fileName: 'Voice message.mp3',
```
Comment (sravan-s) [inline src/smart-components/Channel/context/hooks/useSendVoiceMessageCallback.ts:39]: `Voice message.mp3` are we really gonna use this name? No timestamp? 
PS. make this const also :D

File: src/smart-components/Channel/context/hooks/useSendVoiceMessageCallback.ts [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+      };
+    if (quoteMessage) {
+      messageParams.isReplyToChannel = true;
+      messageParams.parentMessageId = quoteMessage.messageId;
+    }
+    currentGroupChannel.sendFileMessage(messageParams)
+      .onPending((pendingMessage) => {
+        pubSub.publish(topics.SEND_MESSAGE_START, {
+          /* pubSub is used instead of messagesDispatcher
+            to avoid redundantly calling `messageActionTypes.SEND_MESSAGEGE_START` */
```
Comment (sravan-s) [inline src/smart-components/Channel/context/hooks/useSendVoiceMessageCallback.ts:60]: 👍

File: src/smart-components/Channel/context/hooks/useSendVoiceMessageCallback.ts [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+          message: pendingMessage,
+          channel: currentGroupChannel,
+        });
+        setTimeout(() => utils.scrollIntoLast(0, scrollRef), 1000);
+      })
+      .onFailed((err, failedMessage) => {
+        logger.error('Channel: Sending voice message failed!', { failedMessage, err });
+        // // eslint-disable-next-line no-param-reassign
+        // failedMessage.localUrl = URL.createObjectURL(file);
+        // // eslint-disable-next-line no-param-reassign
```
Comment (sravan-s) [inline src/smart-components/Channel/context/hooks/useSendVoiceMessageCallback.ts:70]: maybe remove these?

File: src/ui/Icon/colors.ts [hunk 12]
```diff
@@ -0,0 +1,21 @@
+export const Colors = {
```
Comment (sravan-s) [inline src/ui/Icon/colors.ts:1]: 👍  thanks

File: src/ui/MessageSearchFileItem/index.tsx [hunk 13]
```diff
@@ -23,11 +24,12 @@ export default function MessageSearchFileItem(props: Props): ReactElement {
     onClick,
   } = props;
   const { createdAt, url, name } = message;
-  const fileMessageUrl = url;
   // @ts-ignore
   const sender = message.sender || message._sender;
   const { profileUrl, nickname } = sender;
   const { stringSet, dateLocale } = useLocalization();
+  const isVoiceMsg = isVoiceMessage(message);
+  const displayingFileName = isVoiceMsg ? stringSet.VOICE_MESSAGE : (name || url);
```
Comment (sravan-s) [inline src/ui/MessageSearchFileItem/index.tsx:32]: nit: displayingFileName >  prettyFileName

File: src/ui/VoiceMessageInput/index.scss [hunk 14]
```diff
@@ -0,0 +1,176 @@
...
+@-moz-keyframes flickerAnimation {
+  0%   { opacity:0; }
+  50%  { opacity:1; }
+  100% { opacity:0; }
+}
+@-webkit-keyframes flickerAnimation {
+  0%   { opacity:0; }
+  50%  { opacity:1; }
+  100% { opacity:0; }
+}
```
Comment (sravan-s) [inline src/ui/VoiceMessageInput/index.scss:114]: I think you dont need to do this, there is SCSS auto-prefixer, we can check after release.. Lets make a ticket

File: src/ui/VoiceMessageInput/index.scss [hunk 15]
```diff
@@ -0,0 +1,176 @@
...
+  min-width: 12px;
+  min-height: 12px;
+  border-radius: 50%;
+  @include themed() {
+    background-color: t(error-3);
+  }
+   -webkit-animation: flickerAnimation 1s infinite;
+   -moz-animation: flickerAnimation 1s infinite;
+   -o-animation: flickerAnimation 1s infinite;
+    animation: flickerAnimation 1s infinite;
```
Comment (sravan-s) [inline src/ui/VoiceMessageInput/index.scss:127]: same here, better to use auto-prefix~

**Review by sravan-s — APPROVED** [review]
Thanks @HoonBaek This looks like a gigantic amount of work 🙇

File: src/ui/VoiceMessageInput/index.scss (hunk 14 above)
Comment (sravan-s) [reply src/ui/VoiceMessageInput/index.scss:114]: hmm.. we already do autoprefix
Screen Shot 2023-02-21 at 5 49 43 PM

Screen Shot 2023-02-21 at 5 50 00 PM

File: package.json (hunk 1 above)
Comment (HoonBaek) [reply package.json:54]: Thanks let me follow-up this in the other PR
[UIKIT-3259](https://sendbird.atlassian.net/browse/UIKIT-3259)

File: src/hooks/VoiceRecorder/useVoiceRecorder.tsx (hunk 7 above)
Comment (HoonBaek) [reply src/hooks/VoiceRecorder/useVoiceRecorder.tsx:74]: Thanks so much!

File: src/index.d.ts [hunk 16]
```diff
@@ -849,6 +993,7 @@ interface OpenChannelProviderProps {
   messageLimit?: number;
   onBeforeSendUserMessage?(text: string): UserMessageCreateParams;
   onBeforeSendFileMessage?(file_: File): FileMessageCreateParams;
+  onBeforeSendVoiceMessage?: (file: File, quotedMessage?: UserMessage | FileMessage) => FileMessageCreateParams;
```
Comment (HoonBaek) [inline src/index.d.ts:996]: This shouldn't be in OpenChannelProps @HoonBaek

File: src/hooks/VoicePlayer/voicePlayerEvent.ts (hunk 2 above)
Comment (HoonBaek) [reply src/hooks/VoicePlayer/voicePlayerEvent.ts:38]:
```suggestion
export const generateGroupKey = (channelUrl: string, key: string): string => (`${channelUrl}/${key}`);

```

File: src/hooks/VoiceRecorder/index.tsx (hunk 6 above)
Comment (HoonBaek) [reply src/hooks/VoiceRecorder/index.tsx:101]: https://sendbird.atlassian.net/browse/UIKIT-3260

File: src/smart-components/Channel/context/hooks/useSendVoiceMessageCallback.ts (hunk 11 above)
Comment (HoonBaek) [reply src/smart-components/Channel/context/hooks/useSendVoiceMessageCallback.ts:70]: Yea I checked that user can download the file from the failed file message without this logic

File: src/hooks/VoicePlayer/index.tsx (hunk 4 above)
Comment (HoonBaek) [reply src/hooks/VoicePlayer/index.tsx:41]: https://sendbird.atlassian.net/browse/UIKIT-3260

**Pull request merged**